Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1073 @@
+    @inlinable
+    public static func makeSequence(
+        of elementType: Element.Type = Element.self,
+        backPressureStrategy: Strategy,
+        delegate: Delegate
+    ) -> NewSequence {
+        let sequence = Self(
+            backPressureStrategy: backPressureStrategy,
+            delegate: delegate
+        )
+        let source = Source(storage: sequence.storage)
+
+        return .init(source: source, sequence: sequence)
+    }
+
+    @inlinable
+    /* private */ internal init(
+        backPressureStrategy: Strategy,
+        delegate: Delegate
+    ) {
+        let storage = Storage(
+            backPressureStrategy: backPressureStrategy,
+            delegate: delegate
+        )
+        self.internalClass = .init(storage: storage)
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncSequenceProducer: AsyncSequence {
+    public func makeAsyncIterator() -> AsyncIterator {
+        AsyncIterator(storage: self.internalClass.storage)
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncSequenceProducer {
+    public struct AsyncIterator: AsyncIteratorProtocol {
+        /// This class is needed to hook the deinit to observe once all references to an instance of the ``AsyncIterator`` are dropped.
+        ///
+        /// If we get move-only types we should be able to drop this class and use the `deinit` of the ``AsyncIterator`` struct itself.
+        ///
+        /// - Important: This is safe to be unchecked ``Sendable`` since the `storage` is ``Sendable`` and `immutable`.
+        @usableFromInline
+        /* private */ internal final class InternalClass: @unchecked Sendable {
+            @usableFromInline
+            /* private */ internal let storage: Storage
+
+            fileprivate init(storage: Storage) {
+                self.storage = storage
+                self.storage.iteratorInitialized()
+            }
+
+            @inlinable
+            deinit {
+                self.storage.iteratorDeinitialized()
+            }
+
+            @inlinable
+            /* fileprivate */ internal func next() async -> Element? {
+                await self.storage.next()
+            }
+        }
+
+        @usableFromInline
+        /* private */ internal let internalClass: InternalClass
+
+        fileprivate init(storage: Storage) {
+            self.internalClass = InternalClass(storage: storage)
+        }
+
+        @inlinable
+        public func next() async -> Element? {
+            await self.internalClass.next()
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncSequenceProducer {
+    /// A struct to interface between the synchronous code of the producer and the asynchronous consumer.
+    /// This type allows the producer to synchronously `yield` new elements to the ``NIOAsyncSequenceProducer``
+    /// and to `finish` the sequence.
+    public struct Source {
```

Comment:
Why is this a `struct`? That object has reference semantics, with sideeffects into other threads. I'm very much for using structs as much as we can, but to me this looks like 🐷💄 😅